### PR TITLE
#13-add train and eval loop

### PR DESCRIPTION
**Pull request by c-schicho**

### CODE REVIEW

File: train/train.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    progress_bar = tqdm(range(1, config.num_iter + 1), initial=1, dynamic_ncols=True)
+
+    for n_iter in progress_bar:
+        idx_iter = n_iter - 1
+        if n_iter == 1 or idx_iter in config.milestones:
+            end_iter = config.milestones[n_milestone] if n_milestone < len(config.milestones) else config.num_iter
+            start_iter = config.milestones[n_milestone - 1] if n_milestone > 0 else 0
+            length = config.batch_size[n_milestone] * (end_iter - start_iter)
+            train_dataset = None  # TODO
+            train_loader = None  # TODO
```
Comment (c-schicho) [inline train/train.py:35]: todo after #17 is merged

File: train/eval.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    model = AOSRestoration.get_model_from_config(config)
+    eval_model(model, config)
+
+
+def eval_model(model: AOSRestoration, config: Config) -> Tuple[float, float]:
+    total_psnr = 0.0
+    total_ssim = 0.0
+    n_iter = 0
+
+    data_loader = get_test_loader(batch_size=config.test_batch_size)  # TODO
```
Comment (c-schicho) [inline train/eval.py:21]: todo after #17 is merged

File: train/train.py (hunk 1 above)
Comment (SchmidtPhilipp) [reply train/train.py:35]: Please use:

train_loader, test_loader = data.get_aos_loaders(train_ratio=0.9, train_batch_size = 32, test_batch_size = 32)

or

# choose data sets to include in processing
G_DATASETS = {1: True,
              2: True,
              3: False}

# choose data set locations
G_DATASETS_LOCATION = {1: "../dataset/Part1",
                       2: "../dataset/Part2",
                       3: "../dataset/Part3"}

G_DATASET_LOCATION = [G_DATASETS_LOCATION[dataset_id] for dataset_id, include_dataset in G_DATASETS.items() if include_dataset]

dataset = data.AOSDataset(G_DATASET_LOCATION,
                          relative_path=True,
                          maximum_datasize=None)
                          
train_loader, test_loader = data.AOSDataloader(dataset, train_ratio=0.9)

**Comment by SchmidtPhilipp** [comment]
Please extend the readme with examples especially with training.
Please generalize code with to_device() 
Please test for "cuda" and "mps" available!

**Comment by SchmidtPhilipp** [comment]
Please add info how you split the data and please explain why you do this manually.

**Comment by SchmidtPhilipp** [comment]
Please add devices like: 
device = torch.device('mps' if torch.backends.mps.is_available() else 'cuda' if torch.cuda.is_available() else 'cpu')

**Comment by c-schicho** [comment]
@SchmidtPhilipp 

> Please extend the readme with examples especially for training. Please generalize code with to_device() Please test for "cuda" and "mps" available!

won't add any further information to the readme as this is already contained in the help of the cli, furthermore, we will need to adapt all this anyways to include the model adoption in the later tickets so this information would be outdated right after it is merged. such documentations can be adapted once all of it is fixed (keep in mind everything model and training related is still experimental). 
regarding the mps, since only lena and i will ever train any model this is an unecessary change. also i don't expect any sane person to fully run this script on a cpu (changed it nonetheless). 

> Please add info how you split the data and please explain why you do this manually. Please add more comments.

i just used part 2 of the first simulations as test data but this can be decided by each person individually this is documented for each model individually as i also might change this. i did this beacuse loading it with the existing code includes a random split, however, i need the seed to fix the randomness to compare different settings. but in the end i want to train this models further with either no seed or different seeds to check if we only had a lucky seed or it is true in general. this might not work with a random split in the current implementation because of data leackage (previous test data might now end up in the train set). therefore, using random split would / might just limits us in this regard. 

 i'll just ignore the last comment. as this is not even close to be a constructive comment. imo, the variable names are expressive and the overall structure is pretty basic in this area. furthermore, all special cases are documented with a comment. i could only add redundant information at this point from my point of view.

**Comment by SchmidtPhilipp** [comment]
If you don't want my comments than simply merge it into the main. We do not have to do it in the overcomplicated way. 
BR

**Comment by SchmidtPhilipp** [comment]
Hey @c-schicho, it seems my comments were  little bit too short, and not very specific or accurate, I am sorry. 
Please find the corrected comments below where I already focus on your newer thoughts.

> won't add any further information to the readme as this is already contained in the help of the cli, furthermore, we will need to adapt all this anyways to include the model adoption in the later tickets so this information would be outdated right after it is merged. such documentations can be adapted once all of it is fixed (keep in mind everything model and training related is still experimental).

I think you are in general right, but I just thought as you are providing such a nice README it would be nice to include:
1. an example how to prepare the data: eg:
root
-code
-focal_dataset
-- train
--- set_nr1
-- test
--- set_nr2 


2. how to execute the code, eg.: 
train.py --train True --data_path ../focal_dataset --data_name set_nr1 ...

I mean yes everything is contained in the help but then also the README would be redundant. Still I think for documentation purposes and for the tutors it might be a very nice adoption to include this information, and I do not think that these few lines will change that drastically.

------------------------
> regarding the mps, since only lena and i will ever train any model this is an unecessary change. also i don't expect any sane person to fully run this script on a cpu (changed it nonetheless).

I mean yes many call me in-sane but that cause I am doing far too many courses in parallel. Anyways I do think that in general it makes sense that the code also runs at least on cpu. This would allow me or others to run the code and test other models. But I don't know maybe I am just in-sane for asking for this possibility. ;) 

In regrads to 'mps' that's just the graphical engine of Apple Silicon, it should be compatible* if a proper environment is installed on the machine. 

If you got any concerns about compatibility, I would suggest to either put out a disclaimer that it's either not efficient (cpu) or that it's experimental ('mps'). Another Idea that came to my mind would be to implement it in the parser. But I am sure you having way better ideas :) 

BTW I got it running with 2s/it, not that fast but would be interested how fast you and others can compute. 
*I am still having a fallback to cpu on the sign function of the attention module which is not yet compatible but it is still running through.

------------------------
> i just used part 2 of the first simulations as test data but this can be decided by each person individually this is documented for each model individually as i also might change this. i did this beacuse loading it with the existing code includes a random split, however, i need the seed to fix the randomness to compare different settings. but in the end i want to train this models further with either no seed or different seeds to check if we only had a lucky seed or it is true in general. this might not work with a random split in the current implementation because of data leackage (previous test data might now end up in the train set). therefore, using random split would / might just limits us in this regard.

Ok I think I understood your thoughts. If I understand It right than you want to propose to use all three a test, a validation and a training set. First to separate a test set, that we use only in the end to discriminate between the best implementations,  a validation set and a train set that are used to pick the best model of a distinct architecture and to check the generalization.
What do you think about splitting a test-loader with a fixed seed and using a validation and a train-loader to check if the model generalizes? 


-------------------------

> i'll just ignore the last comment. as this is not even close to be a constructive comment. imo, the variable names are expressive and the overall structure is pretty basic in this area. furthermore, all special cases are documented with a comment. i could only add redundant information at this point from my point of view.

Yes you are totally right, wasn't constructive at all, don't know what I'll thought there. Please just ignore. Sorry.

I wish a Merry Christmas and a Happy New Year!

**Comment by c-schicho** [comment]
@SchmidtPhilipp 

added further information to the readme.  can you please check if it is understandable what to do?

> Anyways I do think that in general it makes sense that the code also runs at least on cpu. This would allow me or others to run the code and test other models. But I don't know maybe I am just in-sane for asking for this possibility. ;)

understood your reasoning, therefore, i also added the support for cpu's. i intentionally wrote *fully* in my initial comment. because this script is limited in terms of using it for smaller testings. as it only saves each 10th batch of images. logging too often may slow down this processes even further, therefore, this decision. also logging all test images would also just increase the tensorboard files too much and finally, i guess no one would check each individual image. so what i actually meant, if you use this script with limited resources, it costs you even more, to produce sample outputs.

> In regrads to 'mps' that's just the graphical engine of Apple Silicon, it should be compatible* if a proper environment is installed on the machine.
> 
> If you got any concerns about compatibility, I would suggest to either put out a disclaimer that it's either not efficient (cpu) or that it's experimental ('mps'). Another Idea that came to my mind would be to implement it in the parser. But I am sure you having way better ideas :)

i simply don't like to push code to a repo, which i cannot even test / verify it's working. changed it nonetheless, since it sounds you already tried it. can you please confirm this assumption?

> BTW I got it running with 2s/it, not that fast but would be interested how fast you and others can compute. *I am still having a fallback to cpu on the sign function of the attention module which is not yet compatible but it is still running through.

did you get this number for training or testing? i got 3-4 it/s during training when i limited the available resources so i could still normally work on my pc. didn't really try to us all the resources during training. as i mentioned once, the actual bottleneck might be my vram. 

> Ok I think I understood your thoughts. If I understand It right than you want to propose to use all three a test, a validation and a training set. First to separate a test set, that we use only in the end to discriminate between the best implementations, a validation set and a train set that are used to pick the best model of a distinct architecture and to check the generalization. What do you think about splitting a test-loader with a fixed seed and using a validation and a train-loader to check if the model generalizes?

actually, i just used test and train data. i kinda introduced a validation set by just limiting the test set to 100 samples during training evaluation. this was simply to speed things up. 

i think in the end you would just like to fix the dataset splitting in some way. as you usually don't want the augmentation to be fixed. i simply decided for the folder approach because it's harder to mess something up. it's easier adoptable for pre-trained models (i could reduce my test size by just moving files and still be sure that i don't leak any data) and the most important one, it was the easiest way to accomplish what i wanted.

back to your question, i think it's also possible to do this in code, since you can provide a generator for the random split. the only downside i see is that i could not adopt the data later, as i'm not aware if we could still guarantee that the new test set is still a sub-set of the original one. but that's just my point of view and i might be overlooking something. what's your take on that one?

> I wish a Merry Christmas and a Happy New Year!

Wish you the same :smile_cat:

**Comment by SchmidtPhilipp** [comment]
> added further information to the readme. can you please check if it is understandable what to do?

Great, love it thanks :) 

> i simply don't like to push code to a repo, which i cannot even test / verify it's working. changed it nonetheless, since it sounds you already tried it. can you please confirm this assumption?

See my suggestions on  13-train_loop_2 I moved the devices to the parser. 
Yes I tried it and it seems to be working well. I need the num_workers=0 but then it is running perfect. 

> did you get this number for training or testing? i got 3-4 it/s during training when i limited the available resources so i could still normally work on my pc. didn't really try to us all the resources during training. as i mentioned once, the actual bottleneck might be my vram.

I got this value training a model. I ran it on the default values, if you did so too then you are about 8 times faster. 

> actually, i just used test and train data. i kinda introduced a validation set by just limiting the test set to 100 samples during training evaluation. this was simply to speed things up.

> i think in the end you would just like to fix the dataset splitting in some way. as you usually don't want the augmentation to be fixed. i simply decided for the folder approach because it's harder to mess something up. it's easier adoptable for pre-trained models (i could reduce my test size by just moving files and still be sure that i don't leak any data) and the most important one, it was the easiest way to accomplish what i wanted.

> back to your question, i think it's also possible to do this in code, since you can provide a generator for the random split. the only downside i see is that i could not adopt the data later, as i'm not aware if we could still guarantee that the new test set is still a sub-set of the original one. but that's just my point of view and i might be overlooking something. what's your take on that one?

I total agree to your toughts! We definitely have to assure that we do not suffer from dataleackage. I am not abgeneigt using the suggested folder solution at all, but I am having doubts that this might not be such easy to reproduce. 
I modified my data loader a bit please have look in 13-train_loop_2. The data is now split according to the data indices in the file names. I also found a bug in the data loader that permitted to load only the first dataset.

----
I made some modifications to your train loop as well please have a look on branch 13-train_loop_2, and tell me what you think about it. :)

**Comment by c-schicho** [comment]
@SchmidtPhilipp 

> See my suggestions on 13-train_loop_2 I moved the devices to the parser. Yes I tried it and it seems to be working well. I need the num_workers=0 but then it is running perfect.

since the functionality is there nonetheless, let's think about this in a follow-up ticket, so we get this finally merged. 

> I got this value training a model. I ran it on the default values, if you did so too then you are about 8 times faster.

was it consistent also for the bigger patch sizes? just out of curiosity.

> I total agree to your thoughts! We definitely have to assure that we do not suffer from dataleackage. I am not abgeneigt using the suggested folder solution at all, but I am having doubts that this might not be such easy to reproduce for others.
> 
> I modified my data loader a bit please have look in 13-train_loop_2. The data is now split according to the data indices in the file names. I also found a bug in the data loader that permitted to load only the first dataset.
> 
> I made some minor modifications to your train loop as well please have a look on branch 13-train_loop_2, and tell me what you think about it. :)

IMO, the "folder"-way it should actually be easier, because you can manually check it and don't rely on the implementation be. as a change in the implementation might cause issues here as it would now. i'm not against your proposal, but if we would change this now, all of my trained models would be worthless, as we could encounter data leakage. therefore, i would go with the folder structure.

had a look at the dataset bug and fixed it with my last commit. also improved the performance with my implementation, it's loading the approx 25k train images in under 10sec now. 

in general, i'm fine with most of the changes but the problem for me is, the branch is broken, therefore, i didn't really test anything or had a deep look.

some problems i saw on the first look: 
- eval loop not working anymore
- arguments were introduced which change nothing
- arguments with wrong default values
- shuffle is gone from the dataloader
- the introduction of an epoch makes no sense in this case

therefore, i would just keep your branch and think about all of it in a follow-up. as i would like to merge this MR today, because i'm already working on the other stuff and it's just annoying to have a lot of sub-branches of one feature branch.

**Review by SchmidtPhilipp — APPROVED** [review]

**Comment by SchmidtPhilipp** [comment]
> since the functionality is there nonetheless, let's think about this in a follow-up ticket, so we get this finally merged.

I totally agree!

> IMO, the "folder"-way it should actually be easier, because you can manually check it and don't rely on the implementation be. as a change in the implementation might cause issues here as it would now. i'm not against your proposal, but if we would change this now, all of my trained models would be worthless, as we could encounter data leakage. therefore, i would go with the folder structure.

Ok, fine for me.
